File: hushh-adk-agents/._kushal_profile_data/Interview-Prep/Problems_By_Pattern/Hash_Map/First_Unique_Character/first_unique_character.js
/**
 * @param {string} s
 * @return {number}
 * 
 * OPTIMIZED APPROACH: Map with Index Tracking
 * - First pass: Build map storing { index, count } for each character
 * - Second pass: Iterate over MAP (at most 26 entries) instead of string (N chars)
 * - Time: O(N) for first pass + O(26) for second pass = O(N)
 * - Space: O(26) = O(1)
 * - Advantage: For very long strings (N >> 26), iterating map is faster than string
 */
var firstUniqChar = function(s) {
    let map = new Map();

    // First pass: Build frequency map with indices
    for(let i = 0; i < s.length; i++)
    {
        if(!map.has(s[i]))
        {
            map.set(s[i], {
                index: i,
                count: 1 
            });
        }
        else
        {
            let obj = map.get(s[i]);
            obj.count = obj.count + 1;
        }
    }

    // Second pass: Iterate over map (26 keys max) to find first unique
    for (const [key, value] of map) {
        if (value.count === 1) {
            return value.index;
        }
    }
    
    return -1;
};

File: hushh-adk-agents/._kushal_profile_data/Interview-Prep/Problems_By_Pattern/Hash_Map/First_Unique_Character/first_unique_character.cs
public class Solution {
    public int FirstUniqChar(string s) {
        // Use an array for O(1) access since we only have lowercase English letters
        int[] count = new int[26];
        
        // First pass: Count frequencies
        foreach (char c in s) {
            count[c - 'a']++;
        }
        
        // Second pass: Find first unique character
        for (int i = 0; i < s.Length; i++) {
            if (count[s[i] - 'a'] == 1) {
                return i;
            }
        }
        
        return -1;
    }
}
